File: core/src/main/java/team/jvav/stuhub/core/data/DAO.java
package team.jvav.stuhub.core.data;

import team.jvav.stuhub.core.data.model.Class;
import team.jvav.stuhub.core.data.model.Group;
import team.jvav.stuhub.core.data.model.Student;

import java.util.ArrayList;
import java.util.List;

/**
 * DAO类，负责管理数据源并提供数据操作接口。
 */
public class DAO {
    /**
     * 存储所有班级信息的列表。
     * <p>
     * 此列表是DAO对数据源的缓存，以避免频繁访问数据源。
     */
    private static final ArrayList<Class> classes = DataSource.loadData();

    /**
     * 保存所有班级信息到数据源。
     *
     * @return 是否成功保存。
     */
    private static boolean saveData() {
        return DataSource.saveData(classes);
    }

    /**
     * 创建一个班级。
     *
     * @param classID 班级ID
     * @return 是否成功创建并保存。当班级已存在时不再重复创建并返回false。
     */
    public static boolean createClass(int classID, String className) {
        for (Class c : classes) {
            if (c.getId() == classID) return false;
        }
        return classes.add(new Class(classID, className)) && saveData();
    }

    /**
     * 删除一个班级。
     *
     * @param classID 班级ID
     * @return 是否成功删除并保存。当班级不存在时不进行任何操作并返回false。
     */
    public static boolean deleteClass(int classID) {
        for (Class c : classes) {
            if (c.getId() == classID) {
                return classes.remove(c) && saveData();
            }
        }
        return false;
    }

    /**
     * 根据班级ID获取班级。
     *
     * @param classID 班级ID
     * @return 班级对象。若不存在则返回null。
     */
    public static Class getClassByID(int classID) {
        for (Class c : classes) {
            if (c.getId() == classID) return c;
        }
        return null;
    }

    /**
     * 获取所有班级。没有班级时返回空列表。
     */
    public static List<Class> getAllClasses() {
        return classes;
    }

    /**
     * 更新一个班级。
     *
     * @param clazz 新的班级对象
     * @return 是否成功更新并保存。当班级不存在时不进行任何操作并返回false。
     */
    public static boolean updateClass(Class clazz) {
        for (int i = 0; i < classes.size(); i++) {
            if (clazz.getId() == classes.get(i).getId()) {
                classes.set(i, clazz);
                return saveData();
            }
        }
        return false;
    }

    /**
     * 在指定班级的指定小组中创建学生。
     *
     * @return 是否成功创建并保存。当班级或小组不存在时不进行任何操作并返回false。
     */
    public static boolean createStudentInGroup(int classID, int groupID, int studentID, String studentName) {
        Class c = getClassByID(classID);
        if (c == null) return false;
        Group g = getGroupByIDInClass(classID, groupID);
        if (g == null) return false;
        return g.addStudent(new Student(studentID, studentName, groupID)) && updateGroupInClass(classID, g);
    }

    /**
     * 删除指定班级的指定小组中的指定学生。
     *
     * @return 是否成功删除并保存。当班级或小组不存在或学生不存在时不进行任何操作并返回false。
     */
    public static boolean deleteStudentInGroup(int classID, int groupID, int studentID) {
        Class c = getClassByID(classID);
        if (c == null) return false;
        Group g = getGroupByIDInClass(classID, groupID);
        if (g == null) return false;
        return g.removeStudent(studentID) && updateGroupInClass(classID, g);
    }

    /**
     * 获取指定班级的指定小组中的指定学生。
     *
     * @return 学生对象或null
     */
    public static Student getStudentInGroup(int classID, int groupID, int studentID) {
        Class c = getClassByID(classID);
        if (c == null) return null;
        Group g = getGroupByIDInClass(classID, groupID);
        if (g == null) return null;
        return g.getStudent(studentID);
    }

    /**
     * 获取指定班级中的指定学生。
     *
     * @return 学生对象或null
     */
    public static Student getStudentInClass(int classID, int studentID) {
        Class c = getClassByID(classID);
        if (c == null) return null;
        for (Group g : c.getAllGroups()) {
            if (g.hasStudent(studentID)) return g.getStudent(studentID);
        }
        return null;
    }

    /**
     * 更新指定班级的指定小组中的指定学生。
     *
     * @return 是否成功更新并保存。当班级或小组不存在或学生不存在时不进行任何操作并返回false。
     */
    public static boolean updateStudentInGroup(int classID, int groupID, int studentId, String studentName) {
        Class c = getClassByID(classID);
        if (c == null) return false;
        Group g = getGroupByIDInClass(classID, groupID);
        if (g == null) return false;
        if (g.hasStudent(studentId)) {
            g.removeStudent(studentId);
            g.addStudent(new Student(studentId, studentName, groupID));
            return updateGroupInClass(classID, g);
        } else return false;
    }

    /**
     * 更新指定班级的的指定学生。
     *
     * @return 是否成功更新并保存。当班级不存在或学生不存在时不进行任何操作并返回false。
     */
    public static boolean updateStudentInClass(int classID, int studentID, String studentName) {
        Class c = getClassByID(classID);
        if (c == null) return false;
        for (Group g : c.getAllGroups()) {
            if (g.hasStudent(studentID)) {
                g.removeStudent(studentID);
                g.addStudent(new Student(studentID, studentName, g.getId()));
                return updateGroupInClass(classID, g);
            }
        }
        System.out.println("Student not found in any group in class " + classID);
        return false;
    }

    /**
     * 将指定班级中指定小组的指定学生移动到另一个小组
     * @return 是否成功移动并保存。当班级或小组不存在或学生不存在或学生已在目标小组时不进行任何操作并返回false。
     */
    public static boolean moveStudentToGroup(int classID, int studentID, int oldGroupID, int newGroupID) {
        if (oldGroupID == newGroupID) {
            System.out.println("oldGroupID and newGroupID cannot be the same");
            return false;
        }
        Student s = getStudentInGroup(classID, oldGroupID, studentID);
        Group oldG = getGroupByIDInClass(classID, oldGroupID);
        Group newG = getGroupByIDInClass(classID, newGroupID);
        if (s == null || oldG == null || newG == null || !oldG.hasStudent(studentID)) {
            System.out.println("Student or group not found or student already in new group");
            return false;
        }
        oldG.removeStudent(studentID);
        newG.addStudent(new Student(studentID, s.getName(), newGroupID));
        return updateGroupInClass(classID, oldG) && updateGroupInClass(classID, newG);
    }

    /**
     * 在指定班级中创建小组。
     *
     * @return 是否成功创建并保存。当班级不存在或小组已存在时不进行任何操作并返回false。
     */
    public static boolean createGroupInClass(int classID, int groupID, String groupName) {
        Class c = getClassByID(classID);
        if (c == null) return false;
        if (c.hasGroup(groupID)) return false;
        return c.addGroup(new Group(groupID, groupName)) && updateClass(c);
    }

    /**
     * 删除指定班级中的小组。
     *
     * @return 是否成功删除并保存。当班级不存在时不进行任何操作并返回false。
     */
    public static boolean deleteGroupInClass(int classID, int groupID) {
        Class c = getClassByID(classID);
        if (c == null) return false;
        return c.removeGroup(groupID) && updateClass(c);
    }

    /**
     * 获取指定班级的指定小组。
     *
     * @return 小组对象。若不存在则返回null。
     */
    public static Group getGroupByIDInClass(int classID, int groupID) {
        Class c = getClassByID(classID);
        if (c == null) return null;
        return c.getGroup(groupID);
    }

    /**
     * 更新指定班级的指定小组。
     *
     * @return 是否成功更新并保存。当班级或小组不存在时不进行任何操作并返回false。
     */
    public static boolean updateGroupInClass(int classID, Group group) {
        Class c = getClassByID(classID);
        if (c == null) return false;
        return c.removeGroup(group.getId()) && c.addGroup(group) && updateClass(c);
    }

    public static ArrayList<Group> getGroupByClassId(int classId) {
        Class c = getClassByID(classId);
        if (c == null) return null;
        return c.getAllGroups();
    }
}

File: core/src/main/java/team/jvav/stuhub/core/data/DAO.kt
package team.jvav.stuhub.core.data

import team.jvav.stuhub.core.data.model.Class
import team.jvav.stuhub.core.data.model.Group
import team.jvav.stuhub.core.data.model.Student

/**
 * DAO类，负责管理数据源并提供数据操作接口。
 */
object DAO {
    /**
     * 存储所有班级信息的列表。
     *
     * 此列表是DAO对数据源的缓存，以避免频繁访问数据源。
     */
    private val classes: MutableList<Class> = DataSource.loadData()

    /**
     * 创建一个班级。若已存在则不重复创建。
     * @param classID 班级ID
     */
    fun createClass(classID: Int) {
        if (classes.any { it.id == classID }) return
        classes.add(Class(classID, mutableListOf(), mutableListOf()))
        DataSource.saveData(classes)
    }

    /**
     * 删除一个班级。若不存在则不进行任何操作。
     * @param classID 班级ID
     */
    fun deleteClass(classID: Int) {
        if (classes.removeIf { it.id == classID }) DataSource.saveData(classes)
    }

    /**
     * 根据班级ID获取班级。
     *
     * @param classID 班级ID
     * @return 班级对象。若不存在则返回null。
     */
    fun getClassByID(classID: Int) = classes.find { it.id == classID }

    /**
     * 获取所有班级列表。
     */
    fun getClasses() = classes.toList()

    /**
     * 更新一个班级。若目前不存在则不进行任何操作。
     * @param clazz 新的班级对象
     */
    fun updateClass(clazz: Class) {
        classes.removeIf { it.id == clazz.id } && classes.add(clazz)
        DataSource.saveData(classes)
    }

    /**
     * 在指定班级中创建一个小组。若班级不存在或小组已存在则不进行任何操作。
     * @param classID 班级ID
     * @param groupID 小组ID
     * @param groupName 小组名称
     */
    fun createGroupInClass(classID: Int, groupID: Int, groupName: String) {
        val clazz = classes.find { it.id == classID } ?: return
        if (clazz.groups.any { it.id == groupID }) return
        val group = Group(groupID, groupName, mutableListOf())
        clazz.groups.add(group)
        DataSource.saveData(classes)
    }

    /**
     * 删除指定班级中的一个小组。若班级或小组不存在则不进行任何操作。
     * @param classID 班级ID
     * @param groupID 小组ID
     */
    fun deleteGroupInClass(classID: Int, groupID: Int) {
        val clazz = classes.find { it.id == classID } ?: return
        clazz.groups.removeIf { it.id == groupID }
        DataSource.saveData(classes)
    }

    /**
     * 获取指定班级中的一个指定小组。
     * @param classID 班级ID
     * @param groupID 小组ID
     * @return 小组对象。若班级或小组不存在则返回null。
     */
    fun getGroupByIDInClass(classID: Int, groupID: Int) =
        classes.find { it.id == classID }?.groups?.find { it.id == groupID }

    /**
     * 获取指定班级中的所有小组列表。
     * @param classID 班级ID
     * @return 小组列表。若班级不存在则返回null。
     */
    fun getGroupsInClass(classID: Int) =
        classes.find { it.id == classID }?.groups?.toList()

    /**
     * 更新指定班级中的一个小组。若班级或小组不存在则不进行任何操作。
     * @param classID 班级ID
     * @param group 新的小组对象
     */
    fun updateGroupInClass(classID: Int, group: Group) {
        val clazz = classes.find { it.id == classID } ?: return
        clazz.groups.removeIf { it.id == group.id } && clazz.groups.add(group)
        DataSource.saveData(classes)
    }

    /**
     * 在指定班级中创建一个学生。若班级不存在或学生已存在则不进行任何操作。
     * @param classID 班级ID
     * @param studentID 学生ID
     * @param studentName 学生名称
     */
    fun createStudentInClass(classID: Int, studentID: Int, studentName: String) {
        val clazz = classes.find { it.id == classID } ?: return
        if (clazz.students.any { it.id == studentID }) return
        clazz.students.add(Student(studentID, studentName))
        DataSource.saveData(classes)
    }

    /**
     * 将指定班级中的指定学生加入到此班级的指定小组。若班级或小组或学生不存在，或学生已在小组中，则不进行任何操作。
     * @param classID 班级ID
     * @param groupID 小组ID
     * @param studentID 学生ID
     */
    fun addStudentToGroup(classID: Int, groupID: Int, studentID: Int) {
        val clazz = classes.find { it.id == classID } ?: return
        val group = clazz.groups.find { it.id == groupID } ?: return
        val student = clazz.students.find { it.id == studentID } ?: return
        if (group.students.any { it.id == studentID }) return
        group.students.add(student)
        updateGroupInClass(classID, group)
    }

    /**
     * 将指定班级中的指定学生从此班级的指定小组中移除。若班级或小组或学生不存在，或学生不在小组中，则不进行任何操作。
     * @param classID 班级ID
     * @param groupID 小组ID
     * @param studentID 学生ID
     */
    fun removeStudentFromGroup(classID: Int, groupID: Int, studentID: Int) {
        val clazz = classes.find { it.id == classID } ?: return
        val group = clazz.groups.find { it.id == groupID } ?: return
        val student = group.students.find { it.id == studentID } ?: return
        if (group.students.remove(student)) updateGroupInClass(classID, group)
    }

    /**
     * 删除指定班级中的一个学生，同时将其从此班级的所有小组中移除。若班级或学生不存在则不进行任何操作。
     * @param classID 班级ID
     * @param studentID 学生ID
     */
    fun deleteStudentInClass(classID: Int, studentID: Int) {
        val clazz = classes.find { it.id == classID } ?: return
        if (clazz.students.removeIf { it.id == studentID }) {
            for (group in clazz.groups) {
                group.students.removeIf { it.id == studentID }
            }
            DataSource.saveData(classes)
        }
    }

    /**
     * 获取指定班级中的所有学生列表。
     * @param classID 班级ID
     * @return 学生列表。若班级不存在则返回null。
     */
    fun getStudentsInClass(classID: Int) = classes.find { it.id == classID }?.students?.toList()

    /**
     * 获取指定班级中的一个指定学生。
     * @param classID 班级ID
     * @param studentID 学生ID
     * @return 学生对象。若班级或学生不存在则返回null。
     */
    fun getStudentByIDInClass(classID: Int, studentID: Int) = getStudentsInClass(classID)?.find { it.id == studentID }

    /**
     * 更新指定班级中的一个学生。若班级或学生不存在则不进行任何操作。
     * @param classID 班级ID
     * @param student 新的学生对象
     */
    fun updateStudentInClass(classID: Int, student: Student) {
        val clazz = classes.find { it.id == classID } ?: return
        clazz.students.removeIf { it.id == student.id } && clazz.students.add(student)
        DataSource.saveData(classes)
    }

    /**
     * 更新指定班级中的指定小组中的一个学生。若班级或小组或学生不存在则不进行任何操作。
     * @param classID 班级ID
     * @param groupID 小组ID
     * @param student 新的学生对象
     */
    fun updateStudentInGroup(classID: Int, groupID: Int, student: Student) {
        val clazz = classes.find { it.id == classID } ?: return
        val group = clazz.groups.find { it.id == groupID } ?: return
        group.students.removeIf { it.id == student.id } && group.students.add(student)
        updateGroupInClass(classID, group)
    }
}